### DISCUSSION TITLE
Autofocus

### DISCUSSION THREAD

**Discussion by blumzi** [blumzi/MAST_unit#1] [Ideas]
Hi Guys,

This will be a discussion about the autofocus feature for the MAST project.  It is dedicated to exchanging ideas, requirements and possibly implementation of the feature.

**Comment by blumzi** [comment]
Additional inputs/outputs to the autofocus feature.

The current PlaneWave API for the feature (as per thePlaneWave PWI4 Programming API (via HTTP) 2022-10-07) presents the following interfaces:
* `/autofocus/start` - starts an autofocus routine
* `/autofocus/stop` - stops an active autofocus routine

The status of a running/done autofocus routine is available via the following status values:
* `autofocus.is_running` - true while autofocus is active, false otherwise
* `autofocus.success` - true if succeeded
* `autofocus.best_position` - on success, best position attained (focuser automatically moves to this position)
* `autofocus.tolerance` - on success, estimate for how far the focuser can move from the best-focus position to produce a 3% increase in the RMS diameter of a star.

To the best of our understanding, the focus is performed over the whole field-of-view acquired by the camera.

The **MAST** project works on one single star at-a-time.  Ideally we would like to focus on this specific star, but can work with being focused on neighboring stars.

### ROI (Region Of Interest)
For the MAST project, we'd be interested in supplying to the autofocus routine a ROI (rectangular in pixel coordinates) which will contain focusing star candidates (**Q**: can/should we supply a list of pixel-coordinates of the actual stars as well?)

What would be the best characteristics of such a ROI in terms of:
* Area size
* Number of focus stars
* Focus stars magnitudes
* Other?

### Failure reason
Could we have some more details about why the autofocus feature failed, so we can better mitigate?

**Comment by blumzi** [comment]
## Autofocus parameters

As per the _PlaneWave PWI4 Programming API (via HTTP) 2022-10-07_, the `autofocus/start` has no autofocusing-related parameters (only a client_id)

From my experience, when the autofocus is started it **_DOES_** read the parameters (_# of images_, _spacing_, _exp. length_, _binning_) from the PWI4 Focus GUI.  

We need to be able to pass these parameters (any others?) via the HTTP request as well.

**Comment by sagiph-weizmann** [comment]
The default should be to perform focus on all stars within the central 200x200 pixels. This is equivalent to 1arcmin - so in most cases we will have a single or a few sources.
In general, we should go to a bright target (magnitude 11-15) to perform telescope focus,
The catalogue of focus targets should be present in each unit PC
